### PR TITLE
Create mutation endpoint to allow upvote/downvote by users

### LINKED ISSUE #53: Create Mutation for User Voting

### PR DESCRIPTION
**Pull request by DanielEFrampton**
### Sidebar Checklist
​
- [x] Ran test suite
- [x] Request reviewers
- [x] Assign yourself and other contributors
- [x] Link to project
- [x] Added to Readme
​
### Issues Resolved
Resolves #53 
​
### Problem Addressed
In order to implement upvote/downvote buttons, we needed the backend to have a GraphQL mutation, tables, and logic to modify the weighted relevancy rating.
​
### Solution Implemented
- ​Created user_votes table which belongs to a method_result
- added a user_score column to method_results
- Created model method on MethodResult which updates the relevancy rating based on percentage of upvotes to total votes
- Created a GraphQL mutation to create a user_vote object and execute that model method on a given method_result and return a fresh top 5 results

### Screenshots (optional)
​

### CODE REVIEW

File: flaskr/app.py [hunk 1]
```diff
@@ -81,16 +81,41 @@ class MethodResult(db.Model):
...
+
+class UserVote(db.Model):
+    __tablename__ = 'user_votes'
+
+    id = db.Column(db.Integer, primary_key=True)
+    method_result_id = db.Column(db.Integer, db.ForeignKey('method_results.id'))
+    type = db.Column(db.String)
+
+    method_result = db.relationship('MethodResult', backref='user_votes')
+
```
Comment (mintona) [inline flaskr/app.py:118]: Wahoo! User votes! This is awesome.

File: flaskr/relevancy_rating_generator.py [hunk 2]
```diff
@@ -4,52 +4,58 @@
...
-    for result in search_results:
-        methods = Method.query.filter_by(language_id=result.target_language_id)
-        for method in methods:
-            method_result = MethodResult(method_id=method.id, search_result_id=result.id)
-            db.session.add(method_result)
+class RelevancyRatingGenerator:
+    def __init__(self):
+        self.js_methods = Method.query.filter_by(language_id=2)
+        self.ruby_methods = Method.query.filter_by(language_id=1)
+
```
Comment (mintona) [inline flaskr/relevancy_rating_generator.py:11]: OOP - very good decision.

File: flaskr/app.py [hunk 3]
```diff
@@ -81,16 +81,41 @@ class MethodResult(db.Model):
     id = db.Column(db.Integer, primary_key=True)
     method_id = db.Column(db.Integer, db.ForeignKey('methods.id'))
     search_result_id = db.Column(db.Integer, db.ForeignKey('search_results.id'))
-    relevance_rating_desciption = db.Column(db.Float)
+    relevance_rating_description = db.Column(db.Float)
     relevance_rating_title = db.Column(db.Float)
     weighted_relevancy_rating = db.Column(db.Float)
+    user_score = db.Column(db.Float)
```
Comment (jobannon) [inline flaskr/app.py:87]: new column time!

File: flaskr/app.py [hunk 4]
```diff
@@ -81,16 +81,41 @@ class MethodResult(db.Model):
...
+    def calc_weighted_relevancy_rating(self):
+        if self.relevance_rating_title >= 0.7:
+            return ((self.relevance_rating_title * 8) + (self.relevance_rating_description * 2)) / 10
+        else:
+            return self.relevance_rating_description
+
+    def update_user_score(self):
+        upvotes = [vote for vote in self.user_votes if vote.type == 'up']
+        score_percentage = ((len(upvotes)) / len(self.user_votes))
+        new_weighted_relevancy_rating = (score_percentage * 0.5) + (self.calc_weighted_relevancy_rating() * 0.5)
```
Comment (jobannon) [inline flaskr/app.py:104]: give user 50 precent weight

File: flaskr/app.py [hunk 5]
```diff
@@ -81,16 +81,41 @@ class MethodResult(db.Model):
...
+        self.weighted_relevancy_rating = new_weighted_relevancy_rating
+        self.user_score = score_percentage
+        db.session.commit()
+
+
+class UserVote(db.Model):
```
Comment (jobannon) [inline flaskr/app.py:110]: new table

File: flaskr/app.py [hunk 6]
```diff
@@ -135,7 +165,31 @@ def resolve_translations(parent, info, method_id, target_language_id):
             order_by(desc(MethodResult.weighted_relevancy_rating)).\
             limit(5).all()
 
-schema = graphene.Schema(query=Query)
+
+class CreateVote(graphene.Mutation):
+    Output = graphene.List(MethodResultObject)
```
Comment (jobannon) [inline flaskr/app.py:170]: per our call... output is "special"

**Pull request merged**